Add close button to Theme Showcase Banner

mendezcode opened the pull request:
This PR adds a close button to the Theme Showcase Banner.

Related: #23944

- - -

Testing Instructions:

1) Go to My Sites > Themes
2) Click on the [ x ] on the Showcase Banner

Expected Behavior: Showcase banner should close, and the closed state should persist across reloads (where the state is not hydrated).

Actual Behavior: Showcase banner closes, but reappears when reloading (when state is not hydrated).

ockham commented on client/my-sites/themes/themes-banner/index.jsx:
```diff
@@ -92,13 +109,18 @@ const mapStateToProps = ( state, { themeId } ) => {
+const mapDispatchToProps = dispatch => ( {
 	recordTracksEvent: recordTracksEventAction,
-} )( localize( ThemesBanner ) );
+	onBannerClose: () => dispatch( hideThemesBanner ),
```
No need to use the `dispatch =>` longform of `mapDispatchToProps`. Let's continue to use the object shorthand instead. Also, we recommend against naming actions (and even component methods) that we pass as event handlers to start with `on...`: rather, they should describe what they do, so let's continue to pass `hideThemesBanner` to the component. Finally, let's continue to inline `mapDispatchToProps`:

```jsx
import { hideThemesBanner as hideThemesBannerAction } from 'state/themes/themes-ui/actions'

/*...*/

export default connect( mapStateToProps, {
	hideThemesBanner: hideThemesBannerAction,
	recordTracksEvent: recordTracksEventAction,
} )( localize( ThemesBanner ) );
```

ockham commented on client/my-sites/themes/themes-banner/style.scss:
```diff
@@ -68,6 +68,28 @@
+			cursor: pointer;
+		}
+
+		svg {
```
I think we should rather select this by its (more specific) `.gridicon-cross-small` (or whateverit's called ) `className`. (I think it's what we generally do in Calypso. In this case, future-proofness is another argument: Imagine we'd add an SVG as a visual element to a banner -- it'd be erroneously selected by an `svg` selector.)

mendezcode replied:
Right, makes total sense indeed!

ockham commented on client/state/action-types.js:
```diff
@@ -983,6 +983,7 @@ export const THEME_UPLOAD_FAILURE = 'THEME_UPLOAD_FAILURE';
 export const THEME_UPLOAD_PROGRESS = 'THEME_UPLOAD_PROGRESS';
 export const THEME_UPLOAD_START = 'THEME_UPLOAD_START';
 export const THEME_UPLOAD_SUCCESS = 'THEME_UPLOAD_SUCCESS';
+export const THEMES_BANNER_IS_SHOWING = 'THEMES_BANNER_IS_SHOWING';
```
An action type should bear a more imperative (or at least verb-like, rather than state-describing) name, so it should probably be `THEMES_BANNER_HIDE`.

ockham replied:
(or `THEMES_BANNER_VISIBILITY_TOGGLE`, if you want to make it more generic.)

ockham replied:
I think `_HIDE` is better tho since it's easier to grok, and we can always add `_SHOW`.

mendezcode replied:
Good idea, done! 👍

ockham commented on client/state/themes/themes-ui/reducer.js:
```diff
@@ -16,4 +16,12 @@ function backPath( state = '/themes', action ) {
+function themesBannerVisibility( state = true, action ) {
+	switch ( action.type ) {
+		case THEMES_BANNER_IS_SHOWING:
+			return action.showing !== undefined ? action.showing : state;
```
Let's simplify to

```js
case THEMES_BANNER_HIDE:
    return false;
```

mendezcode replied:
Done!

mendezcode commented:
Okay, I have refactored the code with the suggestions above, and closing the banner works! The trick was in rewriting the `mapDispatchToProps` and just passing the actions instead. Thanks for that little tip! 😄 

One thing though, the change doesn't persist across reloads. Wondering if I need to do something to trigger saving the state...

ockham commented on client/state/themes/themes-ui/reducer.js:
```diff
@@ -16,4 +16,12 @@ function backPath( state = '/themes', action ) {
+function themesBannerVisibility( state = true, action ) {
+	switch ( action.type ) {
+		case THEMES_BANNER_HIDE:
+			return action.showing !== undefined ? action.showing : state;
```
See https://github.com/Automattic/wp-calypso/pull/24335/files#r182853044 -- let's just `return false;` here :slightly_smiling_face:

ockham commented:
See e.g. https://github.com/Automattic/wp-calypso/blob/a218171bfa8db501a40be9e8bb464e7e29928f75/client/state/themes/schema.js#L57 and how that's used in the corresponding reducer.

ockham commented:
Tests are failing -- let's also add tests for the new sub-reducer, action, and selector (should be fairly simple, since their implementations are not too complex, fortunately).

mendezcode commented:
Fixed and added the tests for the sub-reducers, action & selector.

Now I added some code to persist the banner visibility. It's working across different sections of Calypso, but it's not persisted on reload...

ockham commented on client/state/themes/schema.js:
```diff
@@ -112,3 +112,7 @@ export const themeFiltersSchema = {
 };
+
+export const themesBannerVisibleSchema = {
+	type: [ 'boolean', 'null' ],
```
We shouldn't (need to) allow for `null` :slightly_smiling_face:

ockham replied:
(For the same reason as https://github.com/Automattic/wp-calypso/pull/24335/files#r183109665)

ockham replied:
Let's change this to `type: 'boolean'`

ockham commented on client/state/themes/themes-ui/selectors.js:
```diff
@@ -21,3 +21,12 @@ export function getBackPath( state ) {
+	if ( 'boolean' === typeof themesBannerVisible ) {
+		return themesBannerVisible;
+	}
+	return true;
```
Ooh, that looks complicated :sweat_smile: 

Let's just 

```js
export function isThemesBannerVisible( state ) {
	return state.themes.themesUI.themesBannerVisible;
}
```

:smile: 

Rationale: The reducer hierarchy will be initialized at startup, so it's guaranteed to be there. We're setting the `initialState` to `true`, and when closing the banner, to `false`. So this should guarantee us that `state.themes.themesUI.themesBannerVisible` always holds a valid bool.

mendezcode replied:
Ha! That's exactly what I had before committing it, so I was on the right path!

What you just mentioned aligns with what I was thinking. I'll dig into it and see what may be going on.

ockham commented on client/state/themes/themes-ui/test/actions.js:
```diff
@@ -0,0 +1,23 @@
+	describe( 'hideThemesBanner()', () => {
+		test( 'Should return the expected action object', () => {
+			const action = hideThemesBanner();
+			expect( action ).to.eql( {
```
We try to avoid `chai` in favor of Jest's built-in test helpers now.  Try dropping the `chai` import, an changing this line to `expect( action ).toEqual( {` (see https://facebook.github.io/jest/docs/en/expect.html#toequalvalue).

ockham commented on client/state/themes/themes-ui/test/reducer.js:
```diff
@@ -11,7 +11,13 @@ import { expect } from 'chai';
+	} );
+
+	test( 'should default to a themesBannerVisible of true', () => {
+		expect( state.themesBannerVisible ).to.eql( true );
```
Let's add tests for the `SERIALIZE` and `DESERIALIZE` actions (grep for those in other test files for inspiration). That might also help us discover why we're not successfully persisting state for this reducer.

mendezcode commented:
I have just confirmed that the state is actually being persisted in IndexedDB, but for some reason the persisted state is being set back to the default state (true) when loading the state...

ockham commented on client/my-sites/themes/themes-banner/style.scss:
```diff
@@ -68,6 +68,30 @@
+			cursor: pointer;
+		}
+
+		svg.gridicons-cross-small {
```
No need for the `svg` tag selector -- just go with the class name :slightly_smiling_face:

ockham commented:
(Copying from DM:)

I have one suspicion that it might have to do how we re-hydrate state from what is persisted in the client's IndexedDB vs what we get as initial state from the server (because we're in SSR'd/isomorphic territory here -- there's a chance that this is actually a bug in how we rehydrate).
i.e. somewhere round here https://github.com/Automattic/wp-calypso/blob/3ddf480be671b1c08a2b42c0f6f8c1db4b9a6741/client/state/initial-state.js#L160 the wrong one out of those two dehydrated/persisted states might take precedence

/cc @sirreal @seear for SSR fun times

Anyway, sorry for not getting back to you about this earlier.

However, I think it's in good enough shape to merge as-is and investigate the state persistence issue separately (pending those final two or so comments that I left on the code, that is).

sirreal commented:
I'd love to test and help debug this (https://github.com/Automattic/wp-calypso/pull/24335#issuecomment-383139071 & https://github.com/Automattic/wp-calypso/pull/24335#issuecomment-385930990), but I'm not sure how to jump in. Would you please add some testing instructions to the description?

ockham commented:
Pasting what @mendezcode (per a Slack convo):

![screen shot 2018-05-03 at 1 21 58 pm](https://user-images.githubusercontent.com/96308/39817346-3304dc24-539e-11e8-84bd-622e30548917.png)

So it really looks like persisted state gets wrongly overwritten by re-hydrated state somewhere round https://github.com/Automattic/wp-calypso/blob/3ddf480be671b1c08a2b42c0f6f8c1db4b9a6741/client/state/initial-state.js#L160

Will try to find a fix.

ockham commented:
I think the fix will be to change the order here:

https://github.com/Automattic/wp-calypso/blob/5e1905ab415447b65b8f0f551409a3a5e663a750/client/state/initial-state.js#L94

Not sure if that might cause any unwanted side effects however...

ockham commented on client/my-sites/themes/themes-banner/index.jsx:
```diff
@@ -46,6 +50,12 @@ class ThemesBanner extends PureComponent {
+	// eslint-disable-next-line no-undef
+	handleBannerClose = e => {
+		this.props.hideThemesBanner();
+		e.preventDefault();
```
This doesn't seem to work -- clicking on the X doesn't just close the banner for me, but also redirects me to the theme advertised by the banner.

ockham replied:
This still seems to be happening for me in logged-out mode, but we hopefully fix that separately.

mendezcode commented:
> Not sure if that might cause any unwanted side effects however...

@ockham I have actually tried to fix it that exact same way. It does fix the issue, but some tests do fail. Very minor though, and I think those tests that fail might be worth looking, since they could be alined to the functionality caused by the current implementation.

alisterscott commented:
I had a look at why the e2e tests are failing and it's because they're expecting behaviour that exists on master - so rebasing this branch should make the e2e tests pass against it

mendezcode commented:
@alisterscott Thanks a lot for the heads up!

alisterscott commented:
I tested this and can close the banner. Are we expecting to not show the banner again on refresh? It shows for me on refresh:

![dismiss banner](https://user-images.githubusercontent.com/128826/40707079-f890f528-6432-11e8-8c79-729889275a28.gif)

mendezcode commented:
@alisterscott Yes, the banner should not show again on refresh. There's actually a bug in Calypso that prevents the closed state from persisting: #24758
